### PR TITLE
DE-98/DE-99 - system_setup: Wsl oobe auto reconf mode

### PR DESCRIPTION
**Pull request by patrick330602**
This fixes TODO 2 and TODO 3 (with the last commit): 
2. Implement an API in the server to know if we are in reconfigure or setup mode based on existing UID user >= 1000. Fetch that API from the client and set the variant to wsl_setup or wsl_configuration.
3. In wsl_configuration mode the UI shows ”base”/”advanced” screens. (required models list is already updated automatically)

### CODE REVIEW

File: subiquity/common/apidef.py
Comment (didrocks) [inline subiquity/common/apidef.py:81]: Please add wsl_setup and wsl_reconfigure.

File: subiquity/common/apidef.py
Comment (didrocks) [inline subiquity/common/apidef.py:79]: Please add wsl_setup and wsl_reconfigure in the comment.

File: subiquity/server/server.py
Comment (didrocks) [inline subiquity/server/server.py:121]: I would not store it on the base_model (above), but rather on the application directly. Then, the method is used to adapt the base_model, without storing it.

File: subiquity/models/subiquity.py
Comment (didrocks) [inline subiquity/models/subiquity.py:108]: see comment on server.py. I don’t think we should store the variant here but on the application from client_variant_POST.

File: system_setup/client/client.py
Comment (didrocks) [inline system_setup/client/client.py:45]: We shouldn’t have the Welcome controller loaded in reconfigure mode (see the spec)

File: system_setup/client/client.py
Comment (didrocks) [inline system_setup/client/client.py:42]: We shouldn’t detect reconfiguration mode in the client. The model is basically:

1. Let the server detects based on UID rules if we are in setup or reconfiguration mode.
2. Get it from the server API in the client in which mode we are
3. A little bit silly with the current implementation, but needed… Post the mode back to the server for it to load the expected server models…

Even if this seems counter-intuitive, because of the 3. this allows:
- the server to do the initial detection, as this is where the actual code is running (client could be remote)
- let the client override it if needed (for the tests for instance)

File: system_setup/client/controllers/summary.py
Comment (didrocks) [inline system_setup/client/controllers/summary.py:105]: I think we shuld rather change the summary message without using the user name here. Then the whole variant os.getuid is not needed here as we don’t need real_name.

File: system_setup/client/controllers/wslconfbase.py
Comment (didrocks) [inline system_setup/client/controllers/wslconfbase.py:19]: Same, variant should not be needed in the Base View, it’s the same in both setup and reconfiguration

File: system_setup/common/helpers.py
Comment (didrocks) [inline system_setup/common/helpers.py:29]: Rather than counting the user, you can stop as soon as you have found one matching user and return true

File: system_setup/common/helpers.py
Comment (didrocks) [inline system_setup/common/helpers.py:31]: Return early in the function if is_dryrun_reconfigure is true just below DRYRUN_RECONFIG environment get.

File: system_setup/ui/views/wslconfbase.py
Comment (didrocks) [inline system_setup/ui/views/wslconfbase.py:102]: We should always use that title without using the variant.

**Review by didrocks — CHANGES_REQUESTED** [review]

File: system_setup/ui/views/wslconfbase.py
Comment (patrick330602) [reply system_setup/ui/views/wslconfbase.py:102]: This one is for In wsl_configuration mode the UI shows ”base”/”advanced” screens though, People would be confused that there are two configuration pages with the same title, otherwise we should change the description

File: system_setup/client/controllers/wslconfbase.py
Comment (patrick330602) [reply system_setup/client/controllers/wslconfbase.py:19]: This one is for In wsl_configuration mode the UI shows ”base”/”advanced” screens.

File: system_setup/client/client.py
Comment (patrick330602) [reply system_setup/client/client.py:42]: I am not sure how should I do this since we should then call an async function in __init__, which of course crashes. What is the proper way I should implement this?

File: system_setup/ui/views/wslconfbase.py
Comment (didrocks) [reply system_setup/ui/views/wslconfbase.py:102]: Yes. but my suggestion was to always use that title rather. Those are the base options and can show them as such in setup mode.

File: system_setup/client/controllers/wslconfbase.py
Comment (didrocks) [reply system_setup/client/controllers/wslconfbase.py:19]: See my other comment below, which should make this uneeded :)

File: system_setup/client/client.py
Comment (didrocks) [reply system_setup/client/client.py:42]: This is a good question as this is triggered by the fact of having static class member with the base inherited/+meta class making magic on controllers, replacing the objects. I think of multiple hackish solutions, but maybe @mwhudson or @dbungert has a more upstream-compatible idea?

**Comment by didrocks** [comment]
Something else I didn’t spot yesterday: please target `main` branch instead of `wsl_oobe_reduce_dup`. This has no dependency on `wsl_oobe_reduce_dup` and should not include this commit. As you changed `wsl_oobe_reduce_dup`, you would need otherwise to due a useless rebase (and same later for other branches that depends on this one)

File: system_setup/client/client.py
Comment (mwhudson) [reply system_setup/client/client.py:42]: So I guess there are two broad approaches here.

1. Have all controllers loaded in the client always but have the ones that are inappropriate for the current mode skip themselves (the generic tui client code has support for this, just raise subiquitycore.tuicontroller.Skip from make_ui)
2. Change the set of controllers to be loaded after seeing what the server says

The first is probably easier, tbh, although the second might be aesthetically more pleasing. Nothing in the installer tui client touches the controllers instance until the first status API call has returned so while the ControllerSet class wasn't written with this in mind it looks like assigning `controller_names` between the instance being constructed and calling `load_all` will work fine (another approach would be to construct the ControllerSet later, probably).

Comment (mwhudson) [reply system_setup/client/client.py:42]: WRT the POSTing the variant back... surely something in the system_state server startup code can call set_source_variant on the model during startup?

File: subiquity/common/apidef.py
Comment (patrick330602) [reply subiquity/common/apidef.py]: This is fixed.

File: subiquity/server/server.py
Comment (patrick330602) [reply subiquity/server/server.py]: Updated.

File: system_setup/client/client.py
Comment (patrick330602) [reply system_setup/client/client.py]: Removed.

File: system_setup/client/client.py
Comment (patrick330602) [reply system_setup/client/client.py]: Updated, and tested locally. It should now work properly.

File: system_setup/client/controllers/summary.py
Comment (patrick330602) [reply system_setup/client/controllers/summary.py:105]: The logic is changed for this, needs check whether should be improved.

File: system_setup/ui/views/wslconfbase.py
Comment (patrick330602) [reply system_setup/ui/views/wslconfbase.py]: Addressed

File: system_setup/client/controllers/wslconfbase.py
Comment (patrick330602) [reply system_setup/client/controllers/wslconfbase.py]: addressed

File: system_setup/client/controllers/summary.py
Comment (didrocks) [reply system_setup/client/controllers/summary.py:105]: I don’t see any changes there, please have a look before I make a second review pass.

**Comment by didrocks** [comment]
There is still one non outdated comment that needs to be fixed IMHO and some merge conflicts, do you mind fixing those before requesting a new review?

File: system_setup/client/controllers/summary.py
Comment (patrick330602) [reply system_setup/client/controllers/summary.py]: Sorry for that, fixed now

**Review by didrocks — CHANGES_REQUESTED** [review]
Some changes that were requested in this PR are still not here. Also, the branch needs a rebase and not a merge as already stated.

**Comment by patrick330602** [comment]
@didrocks I just checked and I have fixed everything that was requested, including rebase to the main. Am I missing anything here?

**Comment by didrocks** [comment]
From what I see in a quick check:
- you merged main, not do a `git rebase`
- WSLConfigurationAdvanced still inherites from WSLConfigurationBase (we removed it and asked to remove it multiple times, is the diff not up to date with your local checking? :/)
- There are some comments added like
`# controllers = {
    #    "server": [
`
- In the diff, there are some empty lines added/removed

Please do a rebase, and then do a real serious check that all previously changes are fixed, that there is no leftover of debug and no uneeded diffs.

**Comment by patrick330602** [comment]
> From what I see in a quick check:
> 
> * you merged main, not do a `git rebase`
> * WSLConfigurationAdvanced still inherites from WSLConfigurationBase (we removed it and asked to remove it multiple times, is the diff not up to date with your local checking? :/)
> * There are some comments added like
> 
> ```
> # controllers = {
>     #    "server": [
> ```
> 
> * In the diff, there are some empty lines added/removed
> 
> Please do a rebase, and then do a real serious check that all previously changes are fixed, that there is no leftover of debug and no uneeded diffs.

Updated and fixed. Could you check it again? Thanks!

File: subiquity/client/client.py
Comment (didrocks) [inline subiquity/client/client.py:115]: Remove this ] in diff and keep existing alignement.

File: subiquity/server/server.py
Comment (didrocks) [inline subiquity/server/server.py:254]: This should call `self.set_source_variant()`

File: subiquitycore/core.py
Comment (didrocks) [inline subiquitycore/core.py:36]: Please remove all this file diff, not correct anymore with the current implementation (this was already a comment in the previous review)

File: system_setup/ui/views/summary.py
Comment (didrocks) [inline system_setup/ui/views/summary.py:15]: Remove this and keep existing alignement.

File: system_setup/ui/views/summary.py
Comment (didrocks) [inline system_setup/ui/views/summary.py:63]: Remove this and keep existing alignement.

File: system_setup/ui/views/wslconfbase.py
Comment (didrocks) [inline system_setup/ui/views/wslconfbase.py:100]: Remove this not relevant in the current PR.

**Review by didrocks — CHANGES_REQUESTED** [review]
Some small typos and alignements fixes needed. Then, it will be ready for upstream to review.

**Comment by didrocks** [comment]
You should remove https://github.com/canonical/subiquity/pull/1042/commits/e0884e9943c23c21994c8cc4bedd50f8e87fa612. This commit doesn’t do anything apart from introducing a diff.

**Comment by didrocks** [comment]
You reintroduce the inheritance in https://github.com/canonical/subiquity/pull/1042/commits/675e6afeec3d9e59e194492810cd9752b28cf36c. Please fix this and remove last commit.

**Comment by didrocks** [comment]
Ok, all comments are now resolved, the tests pass. @mwhudson @dbungert, mind having a review of the non system_setup part?

**Review by dbungert — APPROVED** [review]

File: system_setup/server/controllers/configure.py
Comment (mwhudson) [inline system_setup/server/controllers/configure.py:36]: This line isn't necessary, surely?

File: system_setup/client/client.py
Comment (mwhudson) [inline system_setup/client/client.py:51]: Wait, didn't I ask you to parameterize the superclass implementation of this method in a previous PR?

File: subiquity/server/server.py
Comment (mwhudson) [inline subiquity/server/server.py:126]: It would be more consistent for this to call self.app.set_source_variant would it not?

**Review by mwhudson — COMMENTED** [review]
Seems ok, a few questions. Happy for Dan to merge this if you can convince him your responses to my questions are ok :)

File: system_setup/client/client.py
Comment (patrick330602) [reply system_setup/client/client.py:51]: I am not clear about this.... could you elaborate more?

File: system_setup/server/controllers/configure.py
Comment (mwhudson) [reply system_setup/server/controllers/configure.py:36]: Well OK we can argue about how and whether it's needed in that branch then :)

File: system_setup/client/client.py
Comment (mwhudson) [reply system_setup/client/client.py:51]: Having read things a bit more, I don't think you need to add the restart method to this class at all. Can you explain why you did that?

Comment (patrick330602) [reply system_setup/client/client.py:51]: Now I check again, yep you are correct, the restart is not needed here.

File: subiquity/server/server.py
Comment (patrick330602) [reply subiquity/server/server.py]: done for the fixes

File: system_setup/client/client.py
Comment (patrick330602) [reply system_setup/client/client.py]: done for the fixes here.

**Review by mwhudson — APPROVED** [review]
LGTM, thanks for the changes

**Pull request merged**
